### ISSUE TITLE
Add HTTP boot method

### ISSUE THREAD

**Issue by kvaps** [ltsp/ltsp#75]
I'd like to add support for booting via HTTP

We can use it in both cases:
1) for downloading kernel, initramfs and ltsp.conf

   It is working fine for EFI and some new firmwares which implements IPXE bootloader out-of-box. Using HTTP providing better speed and stability it also have no any problems if you going to use ltsp-server behind the NAT.

2) for downloading rootfs image

   This is might be really useful if you going to save client's image fully into RAM #74. This method is free from all the flaws that nfs has, we can simple enable caching, load balancing and other beautiful things which HTTP providing us.

As about implementation we can add support for generate config for [darkhttpd](https://unix4lyfe.org/darkhttpd/) server, this is tiny and very fast http-server for serve static files.

Or consider using pixiecore, amazing project, which can provide everything need (eg. kernel, initramfs and squashed rootfs image) via HTTP.

**Comment by alkisg** [comment]
This too should already be doable with a couple of PRE_INITRD_BOTTOM commands, similar to #74.

I'm not sure where the http: part would need to go in ltsp.ipxe though, for minimal changes in the code base. Here too, patches are welcome. :)

**Comment by kvaps** [comment]
ipxe is supporting http as well, I was using it even with LTSP5.
Thanks for comment, I will think about the implementation!

**Comment by alkisg** [comment]
HTTP booting rpi3 takes 3 seconds, vs 3 minutes over TFTP.
We should definitely support HTTP,
but we should also allow for NFS; if that's enabled in the iPXE build, that's fast too, and it doesn't require an extra service running on the server.

**Comment by alkisg** [comment]
To get HTTP booting, one would get a minimal HTTP server for example by doing:

```shell
cd /srv/ltsp; sudo php -S 0.0.0.0:80
```

And then in /etc/dnsmasq.d/ltsp-dnsmasq.conf, they'd replace `ltsp/ltsp.ipxe` with `http://ltsp/ltsp.ipxe`.

But I think it would be best if we kept ltsp-dnsmasq.conf using TFTP, and put the `chain` command inside ltsp.ipxe itself, like e.g.

```shell
chain ltsp/ltsp.ipxe || goto no-http-continue-here
```

That would allow the client to try http and yet fallback to tftp if it's not available.

**Comment by kvaps** [comment]
Should we us [`BOOT_METHOD`](https://github.com/ltsp/ltsp/pull/51) for configure this booting method?

**Comment by alkisg** [comment]
No this just needs a couple of lines to be changed, either in ltsp-dnsmasq.conf OR at the top of ltsp.ipxe.

All the rest of ltsp.ipxe remains the same as the paths are relative.

**Comment by kvaps** [comment]
> wouldn't need to hardcode the server IP anywhere.

I've done some experiments, and I can say for sure this configuration is working fine for me:

```diff
-dhcp-boot=tag:iPXE,ltsp/ltsp.ipxe
+dhcp-boot=tag:iPXE,"http://${next-server}/ltsp/ltsp.ipxe"
```

> I think it would be best if we kept ltsp-dnsmasq.conf using TFTP

Well, I'd like to not use TFTP anymore :)

**Comment by alkisg** [comment]
> Well, I'd like to not use TFTP anymore :)

How are you downloading undionly.kpxe and snponly.efi? If you have a local ipxe, it's doable, but in *many* cases people will start booting with their normal PXE stacks.

What evaluates ${next-server}, ipxe? This won't work in proxyDHCP setups, see https://github.com/ltsp/ltsp/blob/master/ltsp/server/ipxe/ltsp.ipxe#L12.

**Comment by kvaps** [comment]
> How are you downloading undionly.kpxe and snponly.efi? If you have a local ipxe, it's doable, but in many cases people will start booting with their normal PXE stacks

Yes, my firmware already contains iPXE loader. So it can work without TFTP at all. For whom not having iPXE it will be downloaded over TFTP, that's true. But afterwards it will use the same http address for downloading `ltsp.ipxe` and kernel and initrds

> What evaluates ${next-server}, ipxe? This won't work in proxyDHCP setups, see https://github.com/ltsp/ltsp/blob/master/ltsp/server/ipxe/ltsp.ipxe#L12.

I think we can find some service variable for that. Eg pxelinux have predefined `boot-server` host, I was using it for implementing http booting for LTSP5 https://github.com/kvaps/ltsp5/commit/25395c84c2800b2e06db7232f2cdd1688b8fbe36

> So I mean that supporting HTTP in ltsp-dnsmasq.conf has some benefits and some drawbacks,
but supporting HTTP in ltsp.ipxe also has some benefits and some drawbacks.
>
> In order to support the proxyDHCP setup, I think I prefer the "HTTP in ltsp.ipxe" implementation; except if both options could be properly provided and documented without confusing LTSP administrators too much.

OK, I'm agree, so what about implementing it on both?
- One is optional flag for `ltsp dnsmasq --http=1` https://github.com/ltsp/ltsp/pull/92
- The second one is HTTP and fallback to the TFTP in ipxe script as you mentioned in https://github.com/ltsp/ltsp/issues/75#issuecomment-557529060

**Comment by alkisg** [comment]
> I think we can find some service variable for that.

I was about to say that "it's in 2 different variables so it won't work, it needs an "if" like the one I pasted",
but then I remembered that we do have 2 different lines anyway in ltsp-dnsmasq.conf, one for real and one for proxydhcp,
so it's possible that it will work for proxydhcp too anyway, making it preferable; then we could only implement that one.

> Eg pxelinux have predefined boot-server host

Thanks for that hint, I didn't know it.

**Comment by kvaps** [comment]
I spent some time trying to download image over http over efi and I didn't succeed with that.

Few thoughts about that:

1. Downloading snponly.efi over http is not working for me, I was trying to boot using both of these options:
    ```
    --dhcp-boot=tag:ltsp9,tag:!iPXE,tag:X86-64_EFI,"http://10.36.50.9/ltsp/snponly.efi",ltsp9,10.36.50.9
    --dhcp-boot=tag:ltsp9,tag:!iPXE,tag:X86-64_EFI,"http://10.36.50.9/ltsp/snponly.efi"
    ```
2. Specifying in ltsp.ipxe script:
    ```
    kernel http://10.36.50.9/ltsp/${img}/vmlinuz initrd=ltsp.img initrd=initrd.img ${cmdline}
    initrd http://10.36.50.9/ltsp/ltsp.img
    initrd http://10.36.50.9/ltsp/${img}/initrd.img
    ```
   makes the deal, but booting is hanging on state after all images are successfully downloaded:

```
>> Booting Embedded LOM 1 Port 1 : Mellanox Network Adapter - NIC (PXE IPv4) 

>> Booting PXE over IPv4.
  Station IP address is 10.36.131.11

  Server IP address is 10.36.50.9
  NBP filename is ltsp/snponly.efi
  NBP filesize is 159296 Bytes
 Downloading NBP file...

  NBP file downloaded successfully.
iPXE initialising devices...ok



iPXE 1.0.0+ (a4f8) -- Open Source Network Boot Firmware -- http://ipxe.org
Features: DNS HTTP iSCSI TFTP AoE EFI Menu

net0: f4:03:43:fd:6b:35 using NII on NII-0000:11:00.0 (open)
  [Link:down, TX:0 TXE:0 RX:0 RXE:0]
  [Link status: Unknown (http://ipxe.org/1a086194)]
Configuring (net0 f4:03:43:fd:6b:35)...... ok
net0: 10.36.131.11/255.255.0.0 gw 10.36.0.1
Next server: 10.36.50.9
Filename: ltsp/ltsp.ipxe
tftp://10.36.50.9/ltsp/ltsp.ipxe... ok
ltsp.ipxe : 3009 bytes [script]
http://10.36.50.9/ltsp/x86_64/vmlinuz... ok
http://10.36.50.9/ltsp/ltsp.img... ok
http://10.36.50.9/ltsp/x86_64/initrd.img... ok
```

**Comment by alkisg** [comment]
The main problem with grub is that it doesn't support proxydhcp, so it won't work for most LTSP users.
The people at ipxe are very responsive, maybe you could try to raise this issue in their forums, and they may provide a solution.

Syslinux.efi also had some issues, but it might be a better alternative to grub for now.

**Comment by kvaps** [comment]
Yep, sorry, it was my fault, I forgot to specify serial console kernel options for my clients, and they have no GPU output. That's why I didn't see the output and thought that nothing was stuck.

In the end I find ou, that snponly.efi is working fine but little bit slow. 
So anyway I'm happy to using grub for now .

**Comment by alkisg** [comment]
One way to allow loading the kernel/initrd over http with the current version of LTSP, is to put the following in ltsp.conf:

```shell
[server]
# Use http instead of tftp:
POST_SERVICE_HTTP="python3 -m http.server --directory /srv/tftp &"
POST_DNSMASQ_HTTP="sed 's|ltsp/ltsp.ipxe|http://\${proxydhcp/dhcp-server}:8000/&|' -i /etc/dnsmasq.d/ltsp-dnsmasq.conf && systemctl restart dnsmasq"
```

And restart the involved services:

```shell
systemctl restart ltsp
systemctl restart dnsmasq
```

**Comment by kvaps** [comment]
Short note that `python3 -m http.server` is single-threaded server, better to use something like [darkhttpd](https://unix4lyfe.org/darkhttpd/) or [ltsp-http](https://github.com/ltsp/ltsp/pull/91) feature with nginx or apache server

**Comment by alkisg** [comment]
I'm sure that `python3 -m http.server` isn't a production quality server, but the "single-threaded" part isn't to blame.
Twisted and darkhttpd are single threaded as well.

Anyways, I configured it that way in a school and it seemed to work fine for 10 clients and 20 test reboots; we'll have feedback from the teacher about its stability. :)
In the long term, we'll have #147...

**Comment by kvaps** [comment]
Sorry I was wrong,

```
python2 -m SimpleHTTPServer
```
*\- is single threaded* (supports single connection)

```
python3 -m http.server
```
*\- is multithreaded*  (supports multiple connections)

just tested it like this, servers:

```bash
cd /tmp
truncate -s 100G test
python3 -m http.server
```

```bash
cd /tmp
truncate -s 100G test
python2 -m SimpleHTTPServer
```

clients:
```bash
wget http://localhost:8000/test -O- > /dev/null
```

**Comment by koocotte** [comment]
I just installed webfs (packaged for debian/ubuntu), a light http server, configuration is really fast, a single file in "/etc/webfsd.conf". Default is set to 32 concurrent connections.

I changed the corresponding dhcp-boot to put the full URL, and my download time change from 2min30 to 3-4 sec.

**Comment by alkisg** [comment]
Download time of what? If downloading the kernel and initrd takes more than a minute, then your network has a faulty cable that you need to replace.
iPXE/TFTP/UDP are sensitive to faulty cables and dropped packes, but that also serves as an alert, to fix the network.
Fixing the bad cables/sockets will also make things more stable later on too, after the initial boot.

**Comment by koocotte** [comment]
Hello,

You're right, I mixed two things.
 
On some recent computers, the boot time with the old LTSP5 was painfully slow. Kernel+Initrd take about 2min30. Older computers does not have any problems and boot time is correct. SNMP on the switch does not report any problem, neither the kernel after boot; so it's probably linked to gpxe or the UEFI stack.

ipxe does a great job.
